### PR TITLE
Feature gate for regional PDs

### PR DESCRIPTION
**Pull request by verult**
**What this PR does / why we need it**: Adding beta feature gate around regional PD support.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: Partially addresses #59988

**Special notes for your reviewer**: This feature has already been in alpha for two releases, but at the time it was not gated with a Kubernetes feature gate. Instead it was controlled by a GCE-specific alpha gate. However, there are additional changes with GCE PD StorageClass parameters that we'd like to gate as well, and this is out of scope of GCE alpha gates.

/cc @saad-ali @lavalamp

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/gce_alpha.go
Comment (saad-ali) [inline pkg/cloudprovider/providers/gce/gce_alpha.go]: Double check removing this feature does not cause any failures if old code still passes the flag in.

**Comment by verult** [comment]
/hold
Testing this locally for the time being

File: pkg/cloudprovider/providers/gce/gce_alpha.go
Comment (verult) [reply pkg/cloudprovider/providers/gce/gce_alpha.go]: This is the only spot that passes in an alpha features list that could contain "DiskAlphaAPI", and it just logs the error if the feature doesn't exist.

https://github.com/kubernetes/kubernetes/blob/6a7656b69364337749133f97d7dcb78f63f36b60/pkg/cloudprovider/providers/gce/gce.go#L294

**Comment by verult** [comment]
Manually verified that the feature gate can enable and disable both dynamic provisioning and manual provisioning of PVs associated with regional PDs. Further verified through e2e tests.

/hold cancel

File: pkg/features/kube_features.go
Comment (saad-ali) [inline pkg/features/kube_features.go]: Rename to GCERegionalPersistentDisk

**Comment by verult** [comment]
Addressed comments

File: pkg/cloudprovider/providers/gce/gce_disks.go
Comment (saad-ali) [inline pkg/cloudprovider/providers/gce/gce_disks.go]: Update the error below. It is no longer valid:
`The regional PD feature is only available via the GCE Alpha API. Enable \"DiskAlphaAPI\" in the list of \"alpha-features\" in \"gce.conf\" to use the feature.")`

Ditto below.

Comment (saad-ali) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: Also since you have a k8s feature gate, maybe you can do the gating at the GCE Volume Plugin layer now instead of in the cloud provider code? Your call.

File: pkg/cloudprovider/providers/gce/gce_disks.go
Comment (saad-ali) [inline pkg/cloudprovider/providers/gce/gce_disks.go]: Should we be gating this? At this point the disk has already been created. Doing this will just not report the correct info back to the scheduler. But I can buy the argument, that disabling the feature means disabling *all* of it.

File: pkg/volume/gce_pd/gce_util.go
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_util.go]: Should indicate in error that the feature gate is disabled, and should be enabled if you want to use this feature.

File: pkg/volume/gce_pd/gce_util.go
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_util.go]: nit: swap this statement so termination case is first:
```
if errorCond == true
  return error

thingWeWantToDo()
```

File: pkg/volume/gce_pd/gce_util.go
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_util.go]: What if this `replication-type` is set to `none`? We wouldn't want to fail in that case right.

Why not just put a new check at line 124:

```
if replicationType == replicationTypeRegionalPD 
    && !utilfeature.DefaultFeatureGate.Enabled(features.GCERegionalPersistentDisk) {
  return "errorFeatureNotEnabled"
}
```

Then you can revert all the other changes below.

File: pkg/cloudprovider/providers/gce/gce_disks.go
Comment (verult) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: Yep it's gated at volume plugin layer as well. It's left here as a safeguard.

File: pkg/cloudprovider/providers/gce/gce_disks.go
Comment (verult) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: Without this check, a user can manually provision a PV with with zones label `zone1__zone2` and still have it point to a regional PD. I don't think we should allow this if the flag gate is unset

File: pkg/volume/gce_pd/gce_util.go
Comment (verult) [reply pkg/volume/gce_pd/gce_util.go]: Strictly speaking, `replication-type` doesn't exist before adding regional PD support, so my opinion is `replication-type = none` shouldn't be allowed either.

**Comment by verult** [comment]
/assign @lavalamp @dchen1107 
Need a `pkg/` approver to sign off. Thanks!

File: pkg/cloudprovider/providers/gce/gce_disks.go
Comment (saad-ali) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: Should it fall back to non-regional behavior in that case, or be completely unusable?

Comment (verult) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: The `else` clause here is the non-regional behavior before any regional logic was added

**Comment by verult** [comment]
Addressed ACK'd comments

File: pkg/volume/gce_pd/gce_util.go
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_util.go]: To improve readability, move this to a new method in this file, instead of an inline anonymous function.

File: pkg/volume/gce_pd/gce_util.go
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_util.go]: Is this check really necessary, if you've already gated the possible values of `replicationType` above?

Comment (verult) [reply pkg/volume/gce_pd/gce_util.go]: True. I added the check to be clear exactly which branch the program takes if the gate is set vs unset, but it does hinder readability. Will remove.

**Comment by verult** [comment]
Addressed ACK'd comments

**Review by saad-ali — APPROVED** [review]

**Pull request merged**
